collab-service/src/redis_adapter/redis_pubsub_adapter.ts

```diff
@@ -0,0 +1,82 @@
+import { RedisClientType } from 'redis';
+import Logger from '../utils/logger';
+import {
+  CollabTunnelRequest,
+  CollabTunnelResponse,
+  VerifyRoomErrorCode,
+} from '../proto/collab-service';
+import TunnelPubSub from './redis_pubsub_types';
+
+class RedisPubSubAdapter implements TunnelPubSub<CollabTunnelRequest, CollabTunnelResponse> {
+  redisPub: RedisClientType;
+
+  redisSub: RedisClientType;
+
+  username: string;
+
+  topic: string;
+
+  constructor(
+    redisPub: RedisClientType,
+    redisSub: RedisClientType,
+    username: string,
+    roomId: string,
+  ) {
+    this.redisPub = redisPub;
+    this.redisSub = redisSub;
+    this.username = username;
+    this.topic = roomId;
+  }
+
+  async registerEvent(
+    call: (res: CollabTunnelResponse) => void,
+  ): Promise<void> {
+    await this.redisSub.subscribe(`pubsub-${this.topic}`, (message) => {
+      const messageJson = JSON.parse(message);
+      const {
+        sender,
+        data,
+      } = messageJson;
+      const res = CollabTunnelResponse.create(
+        {
+          data: Buffer.from(data),
+          flags: VerifyRoomErrorCode.VERIFY_ROOM_ERROR_NONE,
+        },
+      );
+      if (sender !== this.username) {
+        call(res);
+      }
+    });
+    Logger.info(`Event ${this.topic} registered by ${this.username}`);
+  }
+
+  async push(request: CollabTunnelRequest): Promise<void> {
+    const messageJson = {
+      sender: this.username,
+      data: request.data,
+    };
+    await this.redisPub.publish(`pubsub-${this.topic}`, JSON.stringify(messageJson));
+  }
```

Comment:
You are assuming Request and Response structure will always be the same, which is not safe.

Reply:
wdym by 'assuming they will always be the same'? is it the `request.data` part?

Reply:
You stringify a Request and parse a Response in the Redis flow.

Reply:
are you suggesting we develop an in-house serialization function then? cuz im not sure how else to pass objects as redis-pubsub expects bytes/strings

Reply:
No, imagine putting an apple into a pipe and expecting an orange to come out the other end. You either put an apple and expect an apple or an orange and expect an orange.

Reply:
so move messageJson stringify thingy out? in a sense, you 'prep' the orange (serialised message) yourself and push() it into the pipe, and on the subscribe() -- you get the orange and unpack it outside as well. keeping to your "Tunnel should be seen as a pipe, without further processing"

Reply:
Yes, Redis is a pipe without further processing. If the Request struct changes your pipe dies under the current way

Reply:
You can push the Response struct directly into Redis thats fine, just don't mix 2 structs.